File: patterns_prac/bt_dp_prac1/fib.py
# Recurrence relation is
# dp[i] = dp[i - 1] + dp[i - 2]


def fib(n: int):
    # Create table where nth fib no
    # is stored
    dp = [0] * (n + 1)
    # update the base case
    dp[0] = 0
    dp[1] = 1
    # iteratively fill the dp table from 2
    for idx in range(2, n + 1):
        dp[idx] = dp[idx - 1] + dp[idx - 2]

    return dp[n]
    # we are returning n here, since dp[0]
    # is not considered at all


print(fib(5))
print(fib(10))

File: patterns_prac/bt_dp_prac2/src/bin/fib.rs
// Recurrence relation
// dp[i] = dp[i - 1] + dp[i - 2]

fn main() {
    println!("{}", fib(5));
    println!("{}", fib(10));
}

fn fib(n: usize) -> i32 {
    // create the table for dp
    let mut dp: Vec<i32> = vec![0; n + 1];
    // update the base cases
    dp[0] = 0;
    dp[1] = 1;
    for idx in 2..n + 1 {
        dp[idx] = dp.get(idx - 1).unwrap() + dp.get(idx - 2).unwrap();
    }
    dp[n]
}
